File: solved/generate-parentheses.ts
/**
 * https://leetcode.com/problems/generate-parentheses/
 * only add parenthese if openCount < n
 * only add parenthese if closedCount < openCount
 * valid if open === closed === n
 */
export function generateParenthesis(n: number): string[] {
  const result: string[] = [];
  const stack: string[] = [];

  function gen(open: number, close: number) {
    if (open === n && close === n) {
      let concatenated = stack.reduce((prev, curr) => prev + curr, '');
      result.push(concatenated);
      return;
    }

    if (open < n) {
      stack.push('(');
      gen(open + 1, close);
      stack.pop();
    }

    if (close < open) {
      stack.push(')');
      gen(open, close + 1);
      stack.pop();
    }
  }

  gen(0, 0);
  return result;
}

File: returned/python/generate_parentheses.py
# https://leetcode.com/problems/generate-parentheses/description/
from typing import List

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []

        def backtrack(current: List[str], open: int, closed: int):
            if len(current) == n*2:
                result.append(''.join(current))
                return
            
            if open < n:
                current.append('(')
                backtrack(current, open+1, closed)
                current.pop()

            if closed < open:
                current.append(')')
                backtrack(current, open, closed+1)
                current.pop()
        
        backtrack([], 0, 0)

        return result